core/shared/src/main/scala/fs2/text.scala

```diff
@@ -35,74 +38,119 @@ object text {
      * 0 is returned.
      */
     def lastIncompleteBytes(bs: Array[Byte]): Int = {
-      val lastThree = bs.drop(0.max(bs.size - 3)).toArray.reverseIterator
-      lastThree
-        .map(continuationBytes)
-        .zipWithIndex
-        .find {
-          case (c, _) => c >= 0
-        }
-        .map {
-          case (c, i) => if (c == i) 0 else i + 1
+      /*
+       * This is logically the same as this
+       * code, but written in a low level way
+       * to avoid any allocations and just do array
+       * access
+       *
+       *
+       *
+        val lastThree = bs.drop(0.max(bs.size - 3)).toArray.reverseIterator
+        lastThree
+          .map(continuationBytes)
+          .zipWithIndex
+          .find {
+            case (c, _) => c >= 0
+          }
+          .map {
+            case (c, i) => if (c == i) 0 else i + 1
+          }
+          .getOrElse(0)
+
+       */
+
+      val minIdx = 0.max(bs.size - 3)
+      var idx = bs.size - 1
+      var counter = 0
+      var res = 0
+      while (minIdx <= idx) {
+        val c = continuationBytes(bs(idx))
+        if (c >= 0) {
+          if (c == counter)
+            res = 0
+          else
+            res = counter + 1
+          // exit the loop
+          idx = 0
         }
-        .getOrElse(0)
+        idx = idx - 1
+        counter = counter + 1
+      }
+      res
     }
 
     def processSingleChunk(
-        outputAndBuffer: (List[String], Chunk[Byte]),
+        bldr: Builder[String, List[String]],
+        buffer: Chunk[Byte],
         nextBytes: Chunk[Byte]
-    ): (List[String], Chunk[Byte]) = {
-      val (output, buffer) = outputAndBuffer
-      val allBytes = Array.concat(buffer.toArray, nextBytes.toArray)
+    ): Chunk[Byte] = {
+      // if processing ASCII or largely ASCII buffer is often empty
+      val allBytes =
+        if (buffer.isEmpty) nextBytes.toArray
+        else Array.concat(buffer.toArray, nextBytes.toArray)
+
       val splitAt = allBytes.size - lastIncompleteBytes(allBytes)
 
-      if (splitAt == allBytes.size)
-        (new String(allBytes.toArray, utf8Charset) :: output, Chunk.empty)
-      else if (splitAt == 0)
-        (output, Chunk.bytes(allBytes))
-      else
-        (
-          new String(allBytes.take(splitAt).toArray, utf8Charset) :: output,
-          Chunk.bytes(allBytes.drop(splitAt))
-        )
+      if (splitAt == allBytes.size) {
```

Comment:
It's a small nit but I'd use `length` instead of `size` when working with arrays:

```scala
scala> val arr = Array(1,2,3)
val arr: Array[Int] = Array(1, 2, 3)

scala> u.reify { arr.size }
val res1: reflect.runtime.universe.Expr[Int] = Expr[Int](Predef.intArrayOps(arr).size)

scala> u.reify { arr.length }
val res2: reflect.runtime.universe.Expr[Int] = Expr[Int](arr.length)
```